ground/src/main/java/com/google/android/ground/persistence/remote/firestore/GeometryConverter.kt

```diff
@@ -76,8 +78,52 @@ class GeometryConverter {
         }
     }
 
-    private fun listToMap(list: List<Any>): Map<Int, Any> {
-        return list.mapIndexedNotNull { index, value -> index to toFirestoreValue(value) }
-            .toMap()
+    private fun indexedMap(list: List<Any>): Map<Int, Any> =
+        list.mapIndexed { index, value -> index to toFirestoreValue(value) }.toMap()
+
+    /**
+     * Converts a `Map` deserialized from Firestore into a `Geometry` instance.
+     */
+    fun fromFirestoreMap(map: Map<String, Any>): Geometry? {
+        val jsonMap = fromFirestoreValue(map)
+        val jsonString = Gson().toJson(jsonMap)
+        val reader = GeoJsonReader()
+        return reader.read(jsonString)
+    }
+
+    private fun fromFirestoreValue(value: Any): Any {
+        return when (value) {
+            is Map<*, *> -> {
+                fromFirestoreValue(value)
+            }
+            is GeoPoint -> {
+                arrayOf(value.latitude, value.longitude)
+            }
+            else -> {
+                value
+            }
+        }
+    }
+
+    private fun fromFirestoreValue(map: Map<*, *>): Any {
+        // If all keys are non-null Ints, assume it refers to an indexed map.
+        // If heuristic breaks, we may also want to check keys are in order starting at 0.
+        return if (map.entries.all { it.key is Int && it.value != null }) {
```

Comment:
NIT: this implementation will need to (presumably) iterate over the complete map multiple times (the call to `all` plus the call to `filter`, plus the call to `map`.

Instead, you could write this as:

```kotlin
return if (map.entries.any { it.key !is Int || it.value == null } ) {
  map.mapNotNull { it.value?.let { fromFirestoreValue(it as Any)} }
} else {
  indexedMapToList(...)...
}
```

Since `map g . filter f === map (g . f)` but it only traverses the structure once, instead of twice. `.any()` is also better than `.all()` since it can return immediately on the first instance of the predicate, rather than traversing the entire structure.

That said, I do think the original impl is more readable, and if the compiler is smart enough it can perform the `filter+map` optimization itself.

Reply:
Feel free to take just a piece of this suggestion too. And, actually the `value?.let` should allow you to drop the cast to any (though it may be dependent on compiler state) Perhaps:

```kotlin
private fun fromFirestoreValue(map: Map<*, *>): Any {
        // If all keys are non-null Ints, assume it refers to an indexed map.
        // If heuristic breaks, we may also want to check keys are in order starting at 0.
        return if (map.entries.all { it.key is Int && it.value != null }) {
          indexedMapToList(map as Map<Int, Any>).map(::fromFirestoreValue)
        } else {
            map.mapNotNull { it.value?.let(::fromFirestoreValue) }
        }
    }
```

would be best.

Reply:
Good catch btw!

Reply:
oh, then this is great! I should've suspected as much.